### PR TITLE
WIP: simple example of adding vendor-specific UAVCAN datatype

### PR DESCRIPTION
**Pull request by kd0aij**
@tridge @EShamaev This places a single vendor specific .uavcan file in  AP_UAVCAN/dsdl/ap_uavcan
and the namespace for the resulting datatype is then ap_uavcan::Example

wscript and uavcangen.py are modified to also parse the files in this directory and the resulting header files are written to libraries/AP_UAVCAN/include/dsdlc_generated in the build folder.

Please let me know if there are any obvious problems with these choices, and advise whether writing a new AP_Example_Backend is the best approach for an example, and if so where it should live.

### CODE REVIEW

**Comment by kd0aij** [comment]
vendor-specific dsdl (only if approved) will reside in a new repo: https://gitter.im/ArduPilot/CANBUS?at=5a046429614889d4755cd29e 
@OXINARF @pavel-kirienko I assume development and testing would occur in the vendor's fork of this repo prior to the (potential) acceptance upstream?

**Comment by pavel-kirienko** [comment]
The namespace name `ap_uavcan` does not adhere to the naming conventions. The proper name would be `ardupilot`, or, in the specific case of Olli's efforts, it could be `storm32`, or some other name that uniquely identifies the project or vendor.

**Comment by kd0aij** [comment]
@pavel-kirienko Would you be willing to create an example vendor-specific dsdl file?
It should be a clear example of what is appropriate for acceptance, and might also contain some examples of what would not be accepted upstream.

**Comment by pavel-kirienko** [comment]
Perhaps, this message would be a more or less acceptable example of a proper vendor-specific data type: https://github.com/kd0aij/ardupilot/blob/f07cce4da7f88d4ea5e17c34870d0bd085360fd9/libraries/AP_UAVCAN/dsdl/uavcan/storm32/8304.Statistics.uavcan (although it could be, perhaps, replaced with a regular standard log message).

All other messages defined in that PR, and the following one in particular, are unlikely to be accepted upstream: <https://github.com/kd0aij/ardupilot/blob/f07cce4da7f88d4ea5e17c34870d0bd085360fd9/libraries/AP_UAVCAN/dsdl/uavcan/storm32/8302.NodeSpecific.uavcan>.

In general, people should keep in mind that UAVCAN is a data bus specification, not yet another bad clone of I2C.

**Comment by kd0aij** [comment]
@pavel-kirienko Thanks, I'll submit a PR with a message similar to that one for use in this PR.
should the datatype ID be 20000?

**Comment by pavel-kirienko** [comment]
> should the datatype ID be 20000?

Yes

**Comment by EShamaev** [comment]
I would advise to making "example" folder where the templates should be put in. There is not need to compile that into codebase.

**Comment by kd0aij** [comment]
@pavel-kirienko This page: http://uavcan.org/Specification/1._Introduction/ states that low latency and high throughput are design goals for uavcan. Are there any quantitative measures of how well it meets those specific goals? 

Also, are there any existing users of the datatype uavcan.equipment.camera_gimbal.AngularCommand?

**Comment by pavel-kirienko** [comment]
> Are there any quantitative measures of how well it meets those specific goals?

Yes. Consider the amount of overhead that UAVCAN adds on top of the application-level traffic. The overhead is very low, therefore applications can exchange more data using fewer CAN frames, which leads to lower latency and higher throughput.

> are there any existing users of the datatype uavcan.equipment.camera_gimbal.AngularCommand?

I am not aware of any. We're going to introduce a few radical changes to several data types in the `uavcan.equipment` namespace once the data type versioning feature is implemented.

**Comment by kd0aij** [comment]
@pavel-kirienko Thanks, but I was looking for actual measurements of latency and throughput in existing implementations. How does the latency for UAVCAN compare to that for one-shot PWM, and what percentage of the bus bandwidth is consumed per ESC at e.g. 400Hz? What effect does other traffic on the bus have on the latency and jitter of ESC control?

**Comment by EShamaev** [comment]
I actually did that research with results and measurement posted either on GitHub or on Gitter. It was more than half year ago so would be hard for me to find now.
The control of ESC over UAVCAN showed good results and was proved parctically.

If the results are needed to be reproduced, that can be done theoretically or practically again without a problem.

**Comment by pavel-kirienko** [comment]
> How does the latency for UAVCAN compare to that for one-shot PWM, and what percentage of the bus bandwidth is consumed per ESC at e.g. 400Hz?

You can easily estimate it. A full CAN frame at 1 Mbps takes about 128 microseconds to propagate. A single CAN frame can carry up to 4 ESC setpoints. You can now easily compare that against one-shot PWM or any other protocol.

CAN FD is much faster than that: a full CAN FD frame takes approximately the same amount of time to propagate while carrying 8 times more data.

> What effect does other traffic on the bus have on the latency and jitter of ESC control?

Other traffic may induce jitter with the uncertainty of up to 148 microseconds @ 1 Mbps (worst case). This uncertainty is introduced when a node is trying to emit a high-priority transfer while the bus is already being occupied by a lower-priority frame. CAN bus guarantees that the jitter/uncertainty will never exceed the worst case duration of a single CAN frame (as long as the bus is functioning correctly).

File: libraries/AP_UAVCAN/AP_UAVCAN.cpp [hunk 1]
```diff
@@ -274,6 +277,25 @@ static void air_data_st_cb1(const uavcan::ReceivedDataStructure<uavcan::equipmen
...
+                data->value = msg.value;
+                data->error_flags = msg.error_flags;
+
+                ap_uavcan->update_example_data((uint16_t) msg.example_id);
+            }
+        }
+    }
+}
+
+X_CREATE_CB_ARR( example_st, ap_uavcan::Example)
```
Comment (magicrub) [inline libraries/AP_UAVCAN/AP_UAVCAN.cpp]: what is X_CREATE_CB_ARR ?

**Comment by magicrub** [comment]
I got a bit lost reading the comment history on this. What is still remaining todo on this PR?

**Comment by kd0aij** [comment]
@magicrub This PR is on hold while I work on a UAVCAN GPS module with Phil Kocmoud.  I don't think I had even decided on what the example message should be yet.  Suggestions appreciated.

**Comment by kd0aij** [comment]
stale

**Pull request closed without merging**